Use reduced range in dynamic linear to avoid instruction overflow

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34692 Use reduced range in dynamic linear to avoid instruction overflow**

We evaluated the accuracy based on the OSS tutorial:
https://pytorch.org/tutorials/intermediate/dynamic_quantization_bert_tutorial.html


Original (no "reduce_range" + per tensor):
 ```
03/08/2020 22:08:20 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:08:20 - INFO - __main__ -     acc = 0.8455882352941176
03/08/2020 22:08:20 - INFO - __main__ -     acc_and_f1 = 0.869494625261272
03/08/2020 22:08:20 - INFO - __main__ -     f1 = 0.8934010152284263
 ```



After the “reduce_range” change ("reduce_range" + per tensor, this PR):
 ```
03/08/2020 22:21:26 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:21:26 - INFO - __main__ -     acc = 0.8480392156862745
03/08/2020 22:21:26 - INFO - __main__ -     acc_and_f1 = 0.8707550030321407
03/08/2020 22:21:26 - INFO - __main__ -     f1 = 0.8934707903780068
 ```


"reduce_range" + per-channel quant:
 ```
03/08/2020 22:33:15 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:33:15 - INFO - __main__ -     acc = 0.8602941176470589
03/08/2020 22:33:15 - INFO - __main__ -     acc_and_f1 = 0.8812619816365654
03/08/2020 22:33:15 - INFO - __main__ -     f1 = 0.902229845626072
 ```


Before "reduce-range" change (no "reduce_range" + per-channel quant, only per channel quant):
```
03/08/2020 22:55:03 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:55:03 - INFO - __main__ -     acc = 0.3161764705882353
03/08/2020 22:55:03 - INFO - __main__ -     acc_and_f1 = 0.15808823529411764
03/08/2020 22:55:03 - INFO - __main__ -     f1 = 0.0
```

Differential Revision: [D20358944](https://our.internmc.facebook.com/intern/diff/D20358944/)

jianyuh committed:
Use reduced range in dynamic linear to avoid instruction overflow

We evaluated the accuracy based on the OSS tutorial:
https://pytorch.org/tutorials/intermediate/dynamic_quantization_bert_tutorial.html


Original (no "reduce_range" + per tensor):
 ```
03/08/2020 22:08:20 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:08:20 - INFO - __main__ -     acc = 0.8455882352941176
03/08/2020 22:08:20 - INFO - __main__ -     acc_and_f1 = 0.869494625261272
03/08/2020 22:08:20 - INFO - __main__ -     f1 = 0.8934010152284263
 ```



After the “reduce_range” change ("reduce_range" + per tensor, this PR):
 ```
03/08/2020 22:21:26 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:21:26 - INFO - __main__ -     acc = 0.8480392156862745
03/08/2020 22:21:26 - INFO - __main__ -     acc_and_f1 = 0.8707550030321407
03/08/2020 22:21:26 - INFO - __main__ -     f1 = 0.8934707903780068
 ```


"reduce_range" + per-channel quant:
 ```
03/08/2020 22:33:15 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:33:15 - INFO - __main__ -     acc = 0.8602941176470589
03/08/2020 22:33:15 - INFO - __main__ -     acc_and_f1 = 0.8812619816365654
03/08/2020 22:33:15 - INFO - __main__ -     f1 = 0.902229845626072
 ```


Before "reduce-range" change (no "reduce_range" + per-channel quant, only per channel quant):
```
03/08/2020 22:55:03 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:55:03 - INFO - __main__ -     acc = 0.3161764705882353
03/08/2020 22:55:03 - INFO - __main__ -     acc_and_f1 = 0.15808823529411764
03/08/2020 22:55:03 - INFO - __main__ -     f1 = 0.0
```

Differential Revision: [D20358944](https://our.internmc.facebook.com/intern/diff/D20358944/)

[ghstack-poisoned]

jianyuh commented:
Addressing #34113

jianyuh committed:
Update on "Use reduced range in dynamic linear to avoid instruction overflow"


We evaluated the accuracy based on the OSS tutorial:
https://pytorch.org/tutorials/intermediate/dynamic_quantization_bert_tutorial.html


Original (no "reduce_range" + per tensor):
 ```
03/08/2020 22:08:20 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:08:20 - INFO - __main__ -     acc = 0.8455882352941176
03/08/2020 22:08:20 - INFO - __main__ -     acc_and_f1 = 0.869494625261272
03/08/2020 22:08:20 - INFO - __main__ -     f1 = 0.8934010152284263
 ```



After the “reduce_range” change ("reduce_range" + per tensor, this PR):
 ```
03/08/2020 22:21:26 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:21:26 - INFO - __main__ -     acc = 0.8480392156862745
03/08/2020 22:21:26 - INFO - __main__ -     acc_and_f1 = 0.8707550030321407
03/08/2020 22:21:26 - INFO - __main__ -     f1 = 0.8934707903780068
 ```


"reduce_range" + per-channel quant:
 ```
03/08/2020 22:33:15 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:33:15 - INFO - __main__ -     acc = 0.8602941176470589
03/08/2020 22:33:15 - INFO - __main__ -     acc_and_f1 = 0.8812619816365654
03/08/2020 22:33:15 - INFO - __main__ -     f1 = 0.902229845626072
 ```


Before "reduce-range" change (no "reduce_range" + per-channel quant, only per channel quant):
```
03/08/2020 22:55:03 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:55:03 - INFO - __main__ -     acc = 0.3161764705882353
03/08/2020 22:55:03 - INFO - __main__ -     acc_and_f1 = 0.15808823529411764
03/08/2020 22:55:03 - INFO - __main__ -     f1 = 0.0
```

Differential Revision: [D20358944](https://our.internmc.facebook.com/intern/diff/D20358944/)

[ghstack-poisoned]

z-a-f commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -70,9 +70,10 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
         /*min=*/x_min,
         /*max=*/x_max,
-        /*qmin=*/is_signed ? -(1 << (precision - 1)) : 0,
+        /*qmin=*/is_signed ? -(1 << (precision - 1 - 1)) : 0,
```
Can you explain why it's `-2`, please?

jianyuh committed:
Update on "Use reduced range in dynamic linear to avoid instruction overflow"


We evaluated the accuracy based on the OSS tutorial:
https://pytorch.org/tutorials/intermediate/dynamic_quantization_bert_tutorial.html


Original (no "reduce_range" + per tensor):
 ```
03/08/2020 22:08:20 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:08:20 - INFO - __main__ -     acc = 0.8455882352941176
03/08/2020 22:08:20 - INFO - __main__ -     acc_and_f1 = 0.869494625261272
03/08/2020 22:08:20 - INFO - __main__ -     f1 = 0.8934010152284263
 ```



After the “reduce_range” change ("reduce_range" + per tensor, this PR):
 ```
03/08/2020 22:21:26 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:21:26 - INFO - __main__ -     acc = 0.8480392156862745
03/08/2020 22:21:26 - INFO - __main__ -     acc_and_f1 = 0.8707550030321407
03/08/2020 22:21:26 - INFO - __main__ -     f1 = 0.8934707903780068
 ```


"reduce_range" + per-channel quant:
 ```
03/08/2020 22:33:15 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:33:15 - INFO - __main__ -     acc = 0.8602941176470589
03/08/2020 22:33:15 - INFO - __main__ -     acc_and_f1 = 0.8812619816365654
03/08/2020 22:33:15 - INFO - __main__ -     f1 = 0.902229845626072
 ```


Before "reduce-range" change (no "reduce_range" + per-channel quant, only per channel quant):
```
03/08/2020 22:55:03 - INFO - __main__ -   ***** Eval results  *****
03/08/2020 22:55:03 - INFO - __main__ -     acc = 0.3161764705882353
03/08/2020 22:55:03 - INFO - __main__ -     acc_and_f1 = 0.15808823529411764
03/08/2020 22:55:03 - INFO - __main__ -     f1 = 0.0
```

Differential Revision: [D20358944](https://our.internmc.facebook.com/intern/diff/D20358944/)

[ghstack-poisoned]

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -1502,7 +1502,7 @@ def test_qlinear(self, batch_size, input_channels, output_channels,
             # Observe X_fp32 and determine X_scale and X_zero_point, this should match
             # internals of dynamic linear.
-            X_scale, X_zp = _calculate_dynamic_qparams(X_fp32, torch.quint8)
+            X_scale, X_zp = _calculate_dynamic_qparams(X_fp32, torch.quint8, reduce_range=True)
```
Please do not use reduce-range as a flag, we are planning to deprecate it. Instead use: 
``` python
          X_scale, X_zp = _calculate_dynamic_qparams(X_fp32, torch.quint8, quant_min=0, quant_max=127)
```

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -70,9 +70,10 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
         /*min=*/x_min,
         /*max=*/x_max,
-        /*qmin=*/is_signed ? -(1 << (precision - 1)) : 0,
+        /*qmin=*/is_signed ? -(1 << (precision - 1 - 1)) : 0,
```
We need to change the API for the op:
```python
      quantized::linear_dynamic(X, W_prepack)
```
to 
``` 
quantized::linear_dynamic(X,W_prepack, quant_min=-128, quant_max=127)
```
And in the dynamic/linear module:
Change forward from:
``` python
   def forward(self, x):
        # Note that we can handle self.bias == None case.
        if self._packed_params.dtype == torch.qint8:
            Y = torch.ops.quantized.linear_dynamic(
                x, self._packed_params._packed_params)
        elif self._packed_params.dtype == torch.float16:
            Y = torch.ops.quantized.linear_dynamic_fp16(
                x, self._packed_params._packed_params)
        else:
            raise RuntimeError('Unsupported dtype on dynamic quantized linear!')
        return Y.to(x.dtype)
```
To
```
   def forward(self, x):
        # Note that we can handle self.bias == None case.
        if self._packed_params.dtype == torch.qint8:
            Y = torch.ops.quantized.linear_dynamic(
                x, self._packed_params._packed_params, quant_min = self.quant_min, quant_max= self.quant_max)
        elif self._packed_params.dtype == torch.float16:
            Y = torch.ops.quantized.linear_dynamic_fp16(
                x, self._packed_params._packed_params)
        else:
            raise RuntimeError('Unsupported dtype on dynamic quantized linear!')
        return Y.to(x.dtype)
```
where we pass quant_min and quant_max as arguments to the init of the module, with default values of 0 and 255.

In
https://github.com/pytorch/pytorch/blob/812b1ad8695c7bf28a8caa2f001f0ec3c554fbb7/torch/nn/quantized/dynamic/modules/linear.py#L85, we do
``` python
        qlinear = Linear(mod.in_features, mod.out_features, dtype=dtype, quant_min=0, quant_max=127)
```
This way, existing models that are already scripted and deployed will not see a change in numerics and newer models can benefit from improved accuracy.

dzhulgakov replied:
Yep, mechanism like that is needed for BC. Please add a test for it too! We can either add explicit range or make it `bool reduce_range`. Explicit range is probably better

raghuramank100 replied:
Yes, there is a PR in the works to remove reduce range and make it explicit

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -70,9 +70,10 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
         /*min=*/x_min,
         /*max=*/x_max,
-        /*qmin=*/is_signed ? -(1 << (precision - 1)) : 0,
+        /*qmin=*/is_signed ? -(1 << (precision - 1 - 1)) : 0,
```
Suggest explicitly overriding quant_min and quant_max as reduce_range is not precise, we are moving to this change across fake-quant and observers.

jianyuh replied:
We also need to consider the changes in QNNPACK path as it also use the same `ChooseQuantizationParams` function.
